diff -up cups-2.3.3op2/backend/lpd.c.lpd-delay cups-2.3.3op2/backend/lpd.c
--- cups-2.3.3op2/backend/lpd.c.lpd-delay	2021-02-01 22:10:25.000000000 +0100
+++ cups-2.3.3op2/backend/lpd.c	2023-06-28 17:28:52.465476261 +0200
@@ -63,7 +63,7 @@ static int	abort_job = 0;		/* Non-zero i
 
 #define RESERVE_NONE		0	/* Don't reserve a priviledged port */
 #define RESERVE_RFC1179		1	/* Reserve port 721-731 */
-#define RESERVE_ANY		2	/* Reserve port 1-1023 */
+#define RESERVE_ANY		2	/* Reserve port 512-1023 */
 
 
 /*
@@ -778,7 +778,7 @@ lpd_queue(const char      *hostname,	/*
 
       if (lport < 721 && reserve == RESERVE_RFC1179)
 	lport = 731;
-      else if (lport < 1)
+      else if (lport < 512)
 	lport = 1023;
 
 #ifdef HAVE_GETEUID
